std/algorithm/sorting.d

```diff
@@ -1048,147 +1048,277 @@ unittest
+                result.source[i] = result.source[i].save;
             }
-            return ret;
+            return result;
         }
     }
 
     static if (allSatisfy!(hasLength, Rs))
     {
-        ///
         @property size_t length()
         {
             size_t result;
-            foreach (i, U; Rs)
+            foreach (i, _; Rs)
             {
-                result += _r[i].length;
+                result += source[i].length;
             }
             return result;
         }
 
-        ///
         alias opDollar = length;
     }
+
+    public Rs source;
+    private size_t _lastFrontIndex = size_t.max;
+    static if (isBidirectional)
+    {
+        private size_t _lastBackIndex = size_t.max;
+    }
 }
 
-/// Ditto
-Merge!(less, Rs) merge(alias less = "a < b", Rs...)
-(Rs rs)
+/**
+   Merge multiple sorted ranges $(D rs) with less-than predicate function $(D
+   pred) into one single sorted output range containing the sorted union of the
+   elements of inputs. Elements in the output are not unique, that is the length
+   of the output is the sum of the lengths of the inputs. (The $(D length)
+   member is offered if all inputs also have $(D length).) The element types of
+   all the inputs must have a common type $(D CommonType).
+
+Params:
+    less = Predicate the given ranges are sorted by.
+    rs = The ranges to compute the union for.
+
+Returns:
+    A range containing the union of the given ranges.
+
+See_Also:
+   $(XREF algorithm, setops, SetUnion)
+
+Details:
+
+All of its inputs are assumed to be sorted. This can mean that inputs are
+   instances of $(XREF range, SortedRange). Use the result of $(XREF algorithm,
+   sort), or $(XREF range, assumeSorted) to merge ranges known to be sorted
+   (show in the example below). However, note that there is currently no way of
+   ensuring that two or more instances of $(XREF range, SortedRange) are sorted
+   using a specific comparison function $(D pred). Therefore no checking is done
+   here to assure that the $(D pred) template parmeter matches the $(D pred)
+   template parameter that were given to each instance of $(XREF range,
+   SortedRange) when the $(D rs) were instantiated.
+
+   This algorithm is lazy, doing work progressively as elements are pulled off
+   the result.
+
+   Average complexity is $(BIGOH n * k) for $(D k) inputs of maximum length $(D n).
+
+   If all inputs have the same element type and offer it by $(D ref), output
+   becomes a range with mutable $(D front) (and $(D back) where appropriate)
+   that reflects in the original inputs.
+
+   If any of the inputs $(D rs) is infinite so is the merge result (`empty` is
+   statically always false).
+
+   Example:
+   -------
+   auto a = [0, 2, 4];
+   auto b = [1, 3, 5];
+   auto c = merge(a.assumeSorted, b.assumeSorted)
+   assert(equal(c, [0, 1, 2, 3, 4, 5]));
+   assert(equal(c.retro, [5, 4, 3, 2, 1, 0]));
+   -------
+*/
+Merge!(less, Rs) merge(alias less = "a < b", Rs...)(Rs rs)
```

Comment:
The constraints should go here, in the user-visible and documented API, not in the `Merge` struct. The constraint should not use the package-level `CommonElementType`.